embassy-nrf/src/uarte.rs

```diff
@@ -281,6 +283,161 @@ impl<'d, T: Instance> Write for Uarte<'d, T> {
     }
 }
 
+/// Interface to an UARTE peripheral that uses timers and PPI to emulate
+/// ReadUntilIdle.
+pub struct UarteWithIdle<'d, U: Instance, T: TimerInstance> {
+    uarte: Uarte<'d, U>,
+    timer: T,
+    _ppi_ch1: Ppi<'d, AnyConfigurableChannel>,
+    _ppi_ch2: Ppi<'d, AnyConfigurableChannel>,
+}
+
+impl<'d, U: Instance, T: TimerInstance> UarteWithIdle<'d, U, T> {
+    /// Creates the interface to a UARTE instance.
+    /// Sets the baud rate, parity and assigns the pins to the UARTE peripheral.
+    ///
+    /// # Safety
+    ///
+    /// The returned API is safe unless you use `mem::forget` (or similar safe mechanisms)
+    /// on stack allocated buffers which which have been passed to [`send()`](Uarte::send)
+    /// or [`receive`](Uarte::receive).
+    #[allow(unused_unsafe)]
+    pub unsafe fn new(
+        uarte: impl Unborrow<Target = U> + 'd,
+        timer: impl Unborrow<Target = T> + 'd,
+        ppi_ch1: impl Unborrow<Target = impl ConfigurableChannel> + 'd,
+        ppi_ch2: impl Unborrow<Target = impl ConfigurableChannel> + 'd,
+        irq: impl Unborrow<Target = U::Interrupt> + 'd,
+        rxd: impl Unborrow<Target = impl GpioPin> + 'd,
+        txd: impl Unborrow<Target = impl GpioPin> + 'd,
+        cts: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        rts: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        config: Config,
+    ) -> Self {
+        let baudrate = config.baudrate;
+        let uarte = Uarte::new(uarte, irq, rxd, txd, cts, rts, config);
+
+        unborrow!(timer, ppi_ch1, ppi_ch2);
+
+        let r = U::regs();
+        let rt = timer.regs();
+
+        // BAUDRATE register values are `baudrate * 2^32 / 16000000`
+        // source: https://devzone.nordicsemi.com/f/nordic-q-a/391/uart-baudrate-register-values
+        //
+        // We want to stop RX if line is idle for 2 bytes worth of time
+        // That is 20 bits (each byte is 1 start bit + 8 data bits + 1 stop bit)
+        // This gives us the amount of 16M ticks for 20 bits.
+        let timeout = 0x8000_0000 / (baudrate as u32 / 40);
+
+        rt.tasks_stop.write(|w| unsafe { w.bits(1) });
+        rt.bitmode.write(|w| w.bitmode()._32bit());
+        rt.prescaler.write(|w| unsafe { w.prescaler().bits(0) });
+        rt.cc[0].write(|w| unsafe { w.bits(timeout) });
+        rt.mode.write(|w| w.mode().timer());
+        rt.shorts.write(|w| {
+            w.compare0_clear().set_bit();
+            w.compare0_stop().set_bit();
+            w
+        });
+
+        let mut ppi_ch1 = Ppi::new(ppi_ch1.degrade_configurable());
+        ppi_ch1.set_event(Event::from_reg(&r.events_rxstarted));
```

Comment:
Maybe I'm just confused about the semantics of this trait. Should it time out if no RX is happening, or should it wait forever?, 

If it should timeout if no RX is happening, then it must use rxstarted I think, because rxdrdy is only raised if it is actually receiving data.

If it should wait forever, it can use rxdrdy, which is only raised once RX is starting to be copied into DMA (at least, this seems to be the behavior from my testing, and from reading infocenter).